DMD.C

```diff
@@ -0,0 +1,489 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2013-2021, Lawrence Livermore National Security, LLC
+ * and other libROM project developers. See the top-level COPYRIGHT
+ * file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ *
+ *****************************************************************************/
+
+// Description: Interface to the DMD algorithm.
+
+#include "DMD.h"
+
+#include "Matrix.h"
+#include "Vector.h"
+#include "scalapack_wrapper.h"
+#include "mpi.h"
+
+/* Use C++11 built-in shared pointers if available; else fallback to Boost. */
+#if __cplusplus >= 201103L
+#include <memory>
+#else
+#include <boost/shared_ptr.hpp>
+#endif
+
+#include "scalapack_wrapper.h"
+
+/* Use automatically detected Fortran name-mangling scheme */
+#define dgeev CAROM_FC_GLOBAL(dgeev, DGEEV)
+#define zgetrf CAROM_FC_GLOBAL(zgetrf, ZGETRF)
+#define zgetri CAROM_FC_GLOBAL(zgetri, ZGETRI)
+
+extern "C" {
+    // Compute eigenvalue and eigenvectors of real non-symmetric matrix.
+    void dgeev(char*, char*, int*, double*, int*, double*, double*, double*, int*, double*, int*, double*, int*, int*);
+
+    // LU decomposition of a general matrix.
+    void zgetrf(int*, int*, double*, int*, int*, int*);
+
+    // Generate inverse of a matrix given its LU decomposition.
+    void zgetri(int*, double*, int*, int*, double*, int*, int*);
+}
+
+using namespace std;
+
+namespace CAROM {
+
+DMD::DMD(int dim)
+{
+    // Get the rank of this process, and the number of processors.
+    int mpi_init;
+    MPI_Initialized(&mpi_init);
+    if (mpi_init == 0) {
+        MPI_Init(nullptr, nullptr);
+    }
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &d_rank);
+    MPI_Comm_size(MPI_COMM_WORLD, &d_num_procs);
+    d_dim = dim;
+}
+
+bool DMD::takeSample(double* u_in)
+{
+    CAROM_VERIFY(u_in != 0);
+
+    Vector sample(u_in, d_dim, true);
+    d_snapshots.push_back(sample);
+    return true;
+}
+
+void DMD::train(double energy_fraction)
+{
+    const Matrix* f_snapshots = getSnapshotMatrix();
+    CAROM_VERIFY(f_snapshots->numColumns() > 1);
+    CAROM_VERIFY(energy_fraction > 0 && energy_fraction <= 1);
+    d_energy_fraction = energy_fraction;
+    d_k = f_snapshots->numColumns() - 1;
+    constructDMD(f_snapshots, d_rank, d_num_procs);
+
+    delete f_snapshots;
+}
+
+void DMD::train(int k)
+{
+    const Matrix* f_snapshots = getSnapshotMatrix();
+    CAROM_VERIFY(f_snapshots->numColumns() > 1);
+    CAROM_VERIFY(k > 0 && k <= f_snapshots->numColumns() - 1);
+    d_energy_fraction = -1;
+    d_k = k;
+    constructDMD(f_snapshots, d_rank, d_num_procs);
+
+    delete f_snapshots;
+}
+
+void
+DMD::constructDMD(const Matrix* f_snapshots,
+                  int d_rank,
+                  int d_num_procs)
+{
+    Matrix* f_snapshots_minus = new Matrix(f_snapshots->numRows(),
```

Comment:
It would be better (and necessary for large problems) to avoid making these two copies of the snapshot matrix, but I realize this may be too much work for this PR. It seems having the copies makes the code simpler, but it is more expensive in terms of memory. This may be fine for the initial implementation, but it should at least be documented that this should be improved in the future for large problems. The amount of work to avoid the copies may be significant. In particular the "minus" snapshot matrix is passed to scalapack, so we would need to extend the scalapack interface to work on a submatrix. Also, we would need `CAROM::Matrix` multiplication for submatrices. It doesn't have to be done in this PR, but some comments would be helpful.

Reply:
Added TODO.